### ISSUE TITLE
Rename to something that will help people know what the hell this is

### ISSUE THREAD

**Issue by GeeH** [zendframework/zend-diactoros#17]
I love clever and unusual names as much as the next man, but this name is just horrid.

**Comment by GeeH** [comment]
The automatic wontfix label makes me :cry:

**Comment by weierophinney** [comment]
Unless you have suggestions, we're not likely to change the name at this point. We couldn't may name it "http" for obvious reasons, "HttpMessage" would never return relevant results, and "Hermes" or "Mercury" are already flooded with results. We briefly considered "zttp" or "zhttp", but ultimately the them out.

The criteria we have:

- cannot conflict with existing ZF component names
- cannot conflict with existing php project/company names
- should have some affinity with either http or messaging
- should be unique enough to rise to the top of search results

I'll gladly consider names that fulfill those criteria, but the name will need to be quite compelling to trigger a rename at this point.

**Comment by GeeH** [comment]
I get that - I'm actually not trying to be argumentative (for once). Personally I just feel that a non-unique descriptive name trumps Googlability.

**Comment by weierophinney** [comment]
It's not just about search; part of this is about being able to track reach. As an example, ZF often appears to trend lower than other frameworks, while simultaneously having more downloads. The reason is interesting: two words in the name, one common, and the other a brand shared amongst several products skews to lower search ranking, which makes understanding our reach difficult. 

Hence, the argument for something more unique. :-)

**Comment by GeeH** [comment]
I get that, it feel so pretentious, I just hate this name.

**Comment by mtymek** [comment]
Agree with GeeH, at the beginning I kept forgetting how it is spelled :) 
Difficult name won't help with getting more reach.

**Comment by GeeH** [comment]
Why can't it just be Zend\Http 3.0?

**Comment by Maks3w** [comment]
`zend-psr7`

**Comment by akrabat** [comment]
Zend\Http clearly doesn't meet the requirements set out above.

**Comment by GeeH** [comment]
Obviously, but my question is why Zend\Http 3.x can't be the PSR-7 implementation? Surely this is exactly the kind of thing that the repositories were split for?

**Comment by weierophinney** [comment]
> why Zend\Http 3.x can't be the PSR-7 implementation? Surely this is exactly the kind of thing that the repositories were split for?

Yes and no.

In this case, the architecture of `Zend\Http` is radically different than a PSR-7 implementation:

- it provides a client implementation
- it provides concrete header implementations
- it provides a cookie-jar implementation

PSR-7 support is around *just* the messages; all of the above are out-of-scope for a PSR-7 implementation. They are instead utilities that would *consume* a PSR-7 implementation at this point

As such, calling this `Zend\Http` 3.0 would mean either (a) dropping all of that support, or require a ton of refactoring/new development to get to feature parity. This would mean we would not have a stable PSR-7 implementation for some months, vs having something that works *immediately*, which is what this library does.

The next consideration is one of migration. The APIs in zend-http and PSR-7 are radically different; in fact, the underlying interfaces themselves have several methods in common, but with differing signatures, making bridging them a bit of a task as well. If we consider that we want to keep the current ZF2 MVC intact for those wanting the more flexible, structured approach it provides, replacing zend-http with this library makes no sense. Those developers then have no migration path, in part because they cannot have a PSR-7 library that can work *in parallel* with zend-http; they have to go all-or-nothing.

By having this library separate from zend-http, we provide a migration path, and we can also defer decisions such as "should ZF offer an HTTP client? should ZF offer concrete header utilities? if so, how?" to later, while providing a PSR-7 library *now*.

**Comment by weierophinney** [comment]
> zend-psr7

No. This is far too generic to be useful when searching for a PSR-7 implementation.

**Comment by weierophinney** [comment]
I'm going to point out one more thing: the fact that the name is currently polarizing also tells me it's likely a good choice. If my tenure doing OSS has taught me anything, any attention is typically good for increasing visibility. :bowtie:

**Comment by Ocramius** [comment]
Not sure why "diactoros" is being bashed this much.
Is a non-english name that hard to remember? (note that I still don't know
how "Gary" is pronounced in welsh :-P )

Marco Pivetta

http://twitter.com/Ocramius

http://ocramius.github.com/

**Comment by GeeH** [comment]
I'm not a fan because it's not immediately obvious what the library does (or even what broader category it belongs in). @weierophinney makes some very good points defending the name, but at the end of the day I hate it.

I'm entitled to an opinion, just as everyone else is entitled to ignore it. At the end of the day, the name isn't actually **that** important.

I prefer it to sit under the `Zend\Http\Message` namespace, or even be called `Zend\HttpMessage`.

In this instance I may be wrong; it's happened once before in 1997.

**Comment by akrabat** [comment]
I'm happy with @weierophinney's reasoning. I just struggle with a name I can't pronounce or spell and hence is easily forgotten. I worry that it limits the chance of it being talked about as a result.

*shrugs*

**Comment by settermjd** [comment]
I appreciate @weierophinney's reasoning as well. It seems sound, and very well thought out. I'd not fully considered the point about traceability in search results, and see that it's definitely easier to know how good the uptake is, when approached from that perspective. On reading that, I'm ok to go along with it. @Ocramius point about "just because it's not English" is a good one as well. That being said, I still find the name quirky, and feel there could be better choices. However, that opinion being voiced, I'm good with the rationale.  Thanks for sharing @weierophinney.

**Comment by harikt** [comment]
Hard to remember, and spell but good to say go to `github.com` `zendframework` repo and look ;) .

**Comment by GeeH** [comment]
Sorry Hari, I have no idea what point you are trying to make.

Gary

**Comment by gianarb** [comment]
Maybe for the future Zend\Http 3.* can use Zend\Diactoros to manage Http\Request and Http\Response and other shared points..
is it useless maintaine two maintain layer? Maybe yes..

**Comment by padraic** [comment]
@GeeH Names don't always refer to what they do very clearly: Symfony, Mockery, Guzzle, Doctrine, Laravel, Twig, Humbug, etc. The only problem, if it even qualifies, is that it's slightly more challenging in the pronunciation sense being derived from Greek versus more familiar English phrases.

**Comment by GeeH** [comment]
A lot of those have meaning, Mockery mocks shit etc.

**Comment by harikt** [comment]
@GeeH what I was trying to say was diactoros is hard to remember name. 

But as we know there is a package in zend framework we can go and look the name when ever needed.

**Comment by GeeH** [comment]
Yeah, it’s fine when your at your desk, but when I am talking to someone in person and I have to say “that package, Dia-something, you’ll have to Google it” - it’s not so good.

G

**Comment by Xerkus** [comment]
Small experiment. I opened those repos last about 4 days ago.
So, we have zend-diactoros(yes, i had too look at address bar to check how to spell it) and that other thing, something like `apigility` but not.
If there will be framework full of those, i will not be happy. Especially if i will need to remember bunch of those namespaces and type them a lot.
That is my only complaint for those catchy names.

**Comment by weierophinney** [comment]
> I still find the name quirky, and feel there could be better choices

As I noted earlier, I'm fine with changing the name if folks can brainstorm ideas that fit the criteria I outlined in my original comment to this issue, and which have general consensus from others. I and others struggled. I really, really wanted "Hermes", which I think would have been quite accepted, but there are already PHP projects with that name, and Zend even has a customer with that name. :-(

**Comment by weierophinney** [comment]
> If there will be framework full of those, i will not be happy. Especially if i will need to remember bunch of those namespaces, read and type them a lot.

Well, the nice part is that you typically only have to type them once per class file, if that. Many times, you'll be type-hinting on the PSR-7 interfaces anyways, which means it will be unusual to need to import the diactoros namespace anyways.

But… noted. We're not planning on renaming existing components, and the two new components were ones we wanted to create differentiation around specifically. I'll keep the creative naming to a minimum.

**Comment by mtymek** [comment]
`zend-diactor`? :-)

**Comment by GeeH** [comment]
Given the way this discussion has gone, I propose Zend\Dictator :D :D :D

**Comment by settermjd** [comment]
In earnest, if the names did carry more cognitive meaning, perhaps rolled off the tongue more easily, and combined innovativeness with descriptiveness, they might gain better traction. Or, perhaps, a theme to run through them, rather like Ubuntu releases. Though this might not be practical. However, I want to give some genuine constructive input.

**Comment by settermjd** [comment]
I'm rather partial to Norse gods myself.

**Comment by padraic** [comment]
Pretty sure zend-harbinger would be free. Just saying. ;) It is actually hard to find words around messages that are not too ordinary or already taken. There's even a PHP-Herald vendor on Packagist.

**Comment by kingharrison** [comment]
What about changing the first letter of what any package does to Z. Easy to remember and figure out what everything does then. Zend Zessage. This would work for everything, trackable, and doesn't require anyone to be clever.

**Comment by kingharrison** [comment]
Going through the list this could probably name all of them uniquely easily and keep with Zend as a brand.
zend-zime
zend-zalidator
zend-zest
zend-zersion
etc....

**Comment by codeliner** [comment]
I think diactoros is a good choice. Like @weierophinney already mentioned it has a high WTF factor. Today I've asked my greek colleage if he know what it means. So people start talking about it :-) 
Just my two cent

**Comment by MarcoDeBortoli** [comment]
I sense Zend-Diactoros will be the new T_PAAMAYIM_NEKUDOTAYIM in PHPub Quizzes' questions, which basically gives all the credit to the reasoning behind its name.
